sys/mips/mips/pm_machdep.c

```diff
@@ -228,35 +249,56 @@ sendsig(sig_t catcher, ksiginfo_t *ksi, sigset_t *mask)
 			sigexit(td, SIGILL);
 			/* NOTREACHED */
 		}
-#endif
+		sp = (uintcap_t)regs->csp;
+#else
 		sp = (vm_offset_t)regs->sp;
+#endif
 	}
-#ifdef CPU_CHERI
-	cp2_len = sizeof(*cfp);
-	sp -= cp2_len;
-	sp &= ~(CHERICAP_SIZE - 1);
-	sf.sf_uc.uc_mcontext.mc_cp2state = sp;
-	sf.sf_uc.uc_mcontext.mc_cp2state_len = cp2_len;
+	sp -= sizeof(*sfp);
+	sp = __builtin_align_down(sp, STACK_ALIGN);
+#if __has_feature(capabilities)
+	sfp = (struct sigframe_c * __capability)sp;
+	sfp = cheri_andperm(sfp, CHERI_CAP_USER_DATA_PERMS);
+#else
+	sfp = (struct sigframe * __capability)sp;
 #endif
-	sp -= sizeof(struct sigframe);
-	sp &= ~(STACK_ALIGN - 1);
-	sfp = (struct sigframe *)sp;
 
 	/* Build the argument list for the signal handler. */
 	regs->a0 = sig;
+#if __has_feature(capabilities)
+	regs->c3 = cheri_csetbounds(&sfp->sf_uc, sizeof(sfp->sf_uc));
+#else
 	regs->a2 = (register_t)(intptr_t)&sfp->sf_uc;
+#endif
 	if (SIGISMEMBER(psp->ps_siginfo, sig)) {
 		/* Signal handler installed with SA_SIGINFO. */
+#if __has_feature(capabilities)
+		regs->c4 = regs->c3;
+		regs->c3 = cheri_csetbounds(&sfp->sf_si, sizeof(sfp->sf_si));
+#else
 		regs->a1 = (register_t)(intptr_t)&sfp->sf_si;
+#endif
 		/* sf.sf_ahu.sf_action = (__siginfohandler_t *)catcher; */
 
 		/* fill siginfo structure */
+#if __has_feature(capabilities)
+		/* XXX: Hack until siginfo_t == siginfo_c */
+		memcpy_c(&sf.sf_si, &ksi->ksi_info, sizeof(sf.sf_si));
+#else
 		siginfo_to_siginfo_native(&ksi->ksi_info, &sf.sf_si);
+#endif
 		sf.sf_si.si_signo = sig;
 	} else {
 		/* Old FreeBSD-style arguments. */
 		regs->a1 = ksi->ksi_code;
+#if __has_feature(capabilities)
+		/*
+		 * XXX: Should this strip tags?
+		 */
+		regs->c4 = ksi->ksi_addr;
+#else
 		regs->a3 = (__cheri_addr uintptr_t)ksi->ksi_addr;
```

Comment:
Why the __cheri_addr in the non-CHERI case?

Reply:
I had left it from the hybrid code, but I've changed it to reduce the diff with upstream.